traffic_portal/app/src/modules/private/cacheStats/CacheStatsController.js

```diff
@@ -17,60 +17,218 @@
-			cacheStatsInterval = undefined;
-		}
-	};
+	/**
+	 * Called by AG-Grid as a pseudo constructor, used when a renderer is
+	 * instantiated.
+	 *
+	 * @param {{
+	 * 	api: any;
+	 * 	colDef: any;
+	 * 	column: any;
+	 * 	columnApi: any;
+	 * 	context: any;
+	 * 	data: CacheStat
+	 * 	eGridCell: HTMLElement;
+	 * 	eParentOfValue: HTMLElement;
+	 * 	formatValue: function;
+	 * 	fullWidth: boolean;
+	 * 	getValue: function;
+	 * 	node: any;
+	 * 	pinned: string | null;
+	 * 	refreshCell: function;
+	 * 	registerRowDragger: function;
+	 * 	rowIndex: number;
+	 * 	setValue: function;
+	 * 	value: boolean;
+	 * 	valueFormatted: null;
+	 * 	$scope: null;
+	 * }} params
+	 */
+	init(params) {
+		this.eGui = document.createElement("span");
+		this.eGui.setAttribute("class", params.value ? "green" : "red");
+		this.eGui.textContent = String(params.value);
+	}
 
-	$scope.cacheStats = cacheStats;
+	/**
+	 * Gets a rendered cell. Parameters are available, but not currently used.
+	 *
+	 * @returns {HTMLElement | Text} A rendered cell element.
+	 */
+	getGui() {
+		return this.eGui;
+	}
+}
 
-	$scope.ssh = serverUtils.ssh;
+/**
+ *
+ * @param {CacheStat[]} cacheStats
+ * @param {{user: {username: string}}} userModel
+ */
+var CacheStatsController = function(cacheStats, $scope, userModel) {
 
-	$scope.bandwidth = function(kbps, unit) {
-		return numberUtils.addCommas(numberUtils.convertTo(kbps, unit));
-	};
+	class CacheStatsSSHCellRenderer {
+		/** @type HTMLAnchorElement | Text */
+		eGui;
 
-	$scope.connections = function(amt) {
-		return numberUtils.addCommas(amt);
-	};
+		/**
+		 * Called by AG-Grid as a pseudo constructor, used when a renderer is
+		 * instantiated.
+		 *
+		 * @param {{
+		 * 	api: any;
+		 * 	colDef: any;
+		 * 	column: any;
+		 * 	columnApi: any;
+		 * 	context: any;
+		 * 	data: CacheStat
+		 * 	eGridCell: HTMLElement;
+		 * 	eParentOfValue: HTMLElement;
+		 * 	formatValue: function;
+		 * 	fullWidth: boolean;
+		 * 	getValue: function;
+		 * 	node: any;
+		 * 	pinned: string | null;
+		 * 	refreshCell: function;
+		 * 	registerRowDragger: function;
+		 * 	rowIndex: number;
+		 * 	setValue: function;
+		 * 	value: string;
+		 * 	valueFormatted: null;
+		 * 	$scope: null;
+		 * }} params
+		 */
+		init(params) {
+			if (params.data.ip === null) {
+				this.eGui = document.createTextNode(params.value);
+				return;
+			}
+			this.eGui = document.createElement("a");
+			this.eGui.href = `ssh://${userModel.user.username}@${params.data.ip}`;
+			this.eGui.setAttribute("class", "link");
+			this.eGui.textContent = params.value;
+		}
 
-	$scope.refresh = function() {
-		$state.reload(); // reloads all the resolves for the view
-	};
+		/**
+		 * Gets a rendered cell. Parameters are available, but not currently used.
+		 *
+		 * @returns {HTMLElement | Text} A rendered cell element.
+		 */
+		getGui() {
+			return this.eGui;
+		}
+	}
 
-	$scope.$on("$destroy", function() {
-		killInterval();
-	});
+	/** The columns of the ag-grid table */
```

Comment:
not very important but this can be a `CGC.ColumnDefinition[]`

Reply:
JSDoc doesn't resolve those references for me, dunno if I'm missing some kind of `@import` or something

Reply:
`/** @typedef { import('./relative/path/to/CommonGridController').CGC } CGC */
` at the top of the file.

Reply:
I tried adding that, it says that `CGC.ColumnDefiniition` is still `any`

Reply:
I added the type anyway